src/Symfony/Component/HttpKernel/EventListener/ErrorListener.php

```diff
@@ -87,9 +88,28 @@ public function onKernelException(ExceptionEvent $event)
         }
     }
 
+    public function onControllerArguments(ControllerArgumentsEvent $event)
+    {
+        $e = $event->getRequest()->attributes->get('exception');
+
+        if (!$e instanceof \Throwable || false === $k = array_search($e, $event->getArguments(), true)) {
+            return;
+        }
+
+        $r = new \ReflectionFunction(\Closure::fromCallable($event->getController()));
+        $r = $r->getParameters()[$k] ?? null;
+
+        if ($r && $r->hasType() && FlattenException::class === $r->getType()->getName()) {
+            $arguments = $event->getArguments();
+            $arguments[$k] = FlattenException::createFromThrowable($e);
+            $event->setArguments($arguments);
+        }
+    }
+
     public static function getSubscribedEvents()
     {
         return [
+            KernelEvents::CONTROLLER_ARGUMENTS => 'onControllerArguments',
```

Comment:
It would be much better if we could add this listener only when necessary, i.e. within the `onKernelException' method.

Reply:
I considered doing this but decided not: mutating the event-dispatcher at runtime is something we'd better avoid all the time. Also, this could introduce subtle order-dependent behaviors.
I think the check is specific enough for having it always enabled, with better unconditional behavior.